packages/astro/src/core/config.ts

```diff
@@ -481,16 +477,35 @@ interface TryLoadConfigResult {
 async function tryLoadConfig(
 	configOptions: LoadConfigOptions,
 	flags: CLIFlags,
-	userConfigPath: string | undefined,
 	root: string
 ): Promise<TryLoadConfigResult | undefined> {
+	let configPath = (await resolveConfigURL({ cwd: configOptions.cwd, flags: configOptions.flags }))
+		?.pathname;
+	if (!configPath) return undefined;
+	let finallyCleanup = async () => {};
 	try {
-		// Automatically load config file using Proload
-		// If `userConfigPath` is `undefined`, Proload will search for `astro.config.[cm]?[jt]s`
+		if (configOptions.invalidateWithCache) {
+			// Hack: Write config to temporary file at project root
+			// This invalidates and reloads file contents when using ESM imports or "resolve"
+			const tempConfigPath = vite.normalizePath(
+				path.join(root, `.temp.${Date.now()}.config${path.extname(configPath)}`)
+			);
+			await fs.promises.writeFile(tempConfigPath, await fs.promises.readFile(configPath));
+			finallyCleanup = async () => {
+				try {
+					await fs.promises.access(tempConfigPath);
+					await fs.promises.unlink(tempConfigPath);
+				} catch {
+					/** file already removed */
+				}
+			};
+			configPath = tempConfigPath;
+		}
+
 		const config = await load('astro', {
-			mustExist: !!userConfigPath,
+			mustExist: true,
```

Comment:
Since we call `resolveConfigURL` at the top now, `undefined` values _should_ be handled before getting to this point. Throw to fallback loader if not.